File: combat.py
"""
V4 combat rules. Substantial rewrite aimed at making mixing emerge.
"""

import random
from dataclasses import dataclass
from enum import Enum

# Advanced Maneuvers
RIPOSTE = False
STOP_HIT = False          # Counter deals extra damage equal to attacker's successes when it wins
DECEPTIVE_ATTACK = False  # Attacker may pump bonus dice into exchange at 2-for-1 reserve cost
EVASIVE_ATTACK = False    # Successful dodger makes a free unopposed attack from remaining reserve

WEAPON_BONUS = 2  # Max damage bonus for attacker; scales +1 per die committed up to this cap
ARMOR_BONUS = 2   # Extra successes for defender on each exchange


class Maneuver(Enum):
    PARRY = "Parry"
    COUNTER = "Counter"
    SIMPLE_ATTACK = "SimpleAttack"
    FEINT = "Feint"
    DODGE = "Dodge"
    DECEPTIVE_ATTACK = "DeceptiveAttack"
    DEFENSELESS = "Defenseless"


@dataclass
class CombatantState:
    max_hd: int
    reserve: int
    exchange: int = 0
    used: int = 0
    lost: int = 0

    @property
    def total_hd(self):
        return self.max_hd - self.lost

    def refresh(self):
        self.reserve = self.total_hd
        self.used = 0
        self.exchange = 0

    def commit(self, dice):
        dice = min(dice, self.reserve)
        self.reserve -= dice
        self.exchange += dice
        return dice
    
    def clear_exchange_to_used(self):
        self.used += self.exchange
        self.exchange = 0
    
    def apply_damage_default(self, damage):
        remaining = damage
        applied = 0
        from_exchange = min(remaining, self.exchange)
        self.exchange -= from_exchange
        self.lost += from_exchange
        remaining -= from_exchange
        applied += from_exchange
        if remaining > 0:
            from_reserve = min(remaining, self.reserve)
            self.reserve -= from_reserve
            self.lost += from_reserve
            remaining -= from_reserve
            applied += from_reserve
        if remaining > 0:
            from_used = min(remaining, self.used)
            self.used -= from_used
            self.lost += from_used
            applied += from_used
        return applied


def roll_successes(num_dice):
    if num_dice <= 0:
        return 0
    return sum(1 for _ in range(num_dice) if random.randint(1, 6) >= 5)


@dataclass
class ExchangeResult:
    attacker_damage_taken: int = 0
    defender_damage_taken: int = 0



def _evasive_attack(dodger, target, evasive_commit):
    if not EVASIVE_ATTACK or evasive_commit <= 0:
        return 0
    committed = dodger.commit(evasive_commit)
    if committed <= 0:
        return 0
    successes = roll_successes(committed)
    damage = max(0, successes + min(WEAPON_BONUS, committed) - ARMOR_BONUS) if successes > 0 else 0
    applied = target.apply_damage_default(damage)
    dodger.clear_exchange_to_used()
    return applied


def _resolve_dodge(attacker, defender, atk_commit, def_commit,
                   atk_maneuver, def_maneuver,
                   atk_followup_commit, def_followup_commit,
                   atk_evasive_commit, def_evasive_commit, result):
    # Phase 1: dodger(s) clear their initial commit to used.
    if atk_maneuver == Maneuver.DODGE:
        attacker.clear_exchange_to_used()
    if def_maneuver == Maneuver.DODGE:
        defender.clear_exchange_to_used()

    # Pairings with no incoming damage either way.
    if atk_maneuver == Maneuver.DODGE and def_maneuver in (Maneuver.DODGE, Maneuver.PARRY):
        if def_maneuver == Maneuver.PARRY:
            defender.clear_exchange_to_used()
        return result
    if def_maneuver == Maneuver.DODGE and atk_maneuver == Maneuver.DODGE:
        return result

    # SA / DA (atk) vs Dodge (def)
    if atk_maneuver in (Maneuver.SIMPLE_ATTACK, Maneuver.DECEPTIVE_ATTACK) and def_maneuver == Maneuver.DODGE:
        atk_successes = roll_successes(attacker.exchange)
        defender.commit(def_followup_commit)
        dodge_roll = roll_successes(defender.exchange)
        if dodge_roll >= max(1, ARMOR_BONUS):
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
            result.attacker_damage_taken += _evasive_attack(defender, attacker, def_evasive_commit)
        else:
            damage = max(0, atk_successes + min(WEAPON_BONUS, attacker.exchange) - ARMOR_BONUS) if atk_successes > 0 else 0
            result.defender_damage_taken = defender.apply_damage_default(damage)
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
        return result

    # Counter (def) vs Dodge (atk)
    if atk_maneuver == Maneuver.DODGE and def_maneuver == Maneuver.COUNTER:
        attacker.commit(atk_followup_commit)
        dodge_roll = roll_successes(attacker.exchange)
        def_successes = roll_successes(def_commit)
        if dodge_roll >= max(1, ARMOR_BONUS):
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
            result.defender_damage_taken += _evasive_attack(attacker, defender, atk_evasive_commit)
        else:
            damage = max(0, def_successes + min(WEAPON_BONUS, defender.exchange) - ARMOR_BONUS) if def_successes > 0 else 0
            result.attacker_damage_taken = attacker.apply_damage_default(damage)
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
        return result

    # Feint (atk) vs Dodge (def)
    if atk_maneuver == Maneuver.FEINT and def_maneuver == Maneuver.DODGE:
        attacker.clear_exchange_to_used()
        followup = min(atk_followup_commit, attacker.reserve)
        attacker.commit(followup)
        followup_successes = roll_successes(followup)
        defender.commit(def_followup_commit)
        dodge_roll = roll_successes(defender.exchange)
        if dodge_roll >= max(1, ARMOR_BONUS):
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
            result.attacker_damage_taken += _evasive_attack(defender, attacker, def_evasive_commit)
        else:
            damage = max(0, followup_successes + min(WEAPON_BONUS, attacker.exchange) - ARMOR_BONUS) if followup_successes > 0 else 0
            result.defender_damage_taken = defender.apply_damage_default(damage)
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
        return result

    return result


def resolve_exchange(attacker, defender, atk_commit, def_commit,
                     atk_maneuver, def_maneuver,
                     atk_followup_commit=0, def_followup_commit=0,
                     atk_bonus=0,
                     atk_evasive_commit=0, def_evasive_commit=0):
    result = ExchangeResult()
    attacker.commit(atk_commit)
    defender.commit(def_commit)

    if def_maneuver == Maneuver.DEFENSELESS:
        def_maneuver = Maneuver.PARRY

    if atk_maneuver == Maneuver.DECEPTIVE_ATTACK and DECEPTIVE_ATTACK:
        bonus = max(0, min(atk_bonus, attacker.reserve // 2))
        attacker.reserve -= 2 * bonus
        attacker.exchange += bonus
        attacker.used += bonus

    if atk_maneuver == Maneuver.DODGE or def_maneuver == Maneuver.DODGE:
        return _resolve_dodge(attacker, defender, atk_commit, def_commit,
                              atk_maneuver, def_maneuver,
                              atk_followup_commit, def_followup_commit,
                              atk_evasive_commit, def_evasive_commit, result)

    if atk_maneuver in (Maneuver.SIMPLE_ATTACK, Maneuver.DECEPTIVE_ATTACK):
        atk_rolled = attacker.exchange
        atk_successes = roll_successes(atk_rolled)

        if def_maneuver == Maneuver.PARRY:
            def_rolled = defender.exchange
            def_successes = roll_successes(def_rolled)
            if atk_successes > def_successes:
                damage = max(0, (atk_successes - def_successes) + min(WEAPON_BONUS, atk_rolled) - ARMOR_BONUS)
            else:
                damage = 0
            result.defender_damage_taken = defender.apply_damage_default(damage)
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
            if RIPOSTE and def_successes > atk_successes:
                riposte_damage = max(0, (def_successes - atk_successes) + min(WEAPON_BONUS, def_rolled) - ARMOR_BONUS)
                if riposte_damage > 0:
                    result.attacker_damage_taken = attacker.apply_damage_default(riposte_damage)

        elif def_maneuver == Maneuver.COUNTER:
            total_atk_dmg = max(0, atk_successes + min(WEAPON_BONUS, atk_rolled) - ARMOR_BONUS) if atk_successes > 0 else 0
            dmg_to_exchange = min(total_atk_dmg, defender.exchange)
            defender.exchange -= dmg_to_exchange
            defender.lost += dmg_to_exchange
            remaining_atk_dmg = total_atk_dmg - dmg_to_exchange
            if remaining_atk_dmg > 0:
                defender.apply_damage_default(remaining_atk_dmg)
            result.defender_damage_taken = atk_successes
            if defender.total_hd > 0 and defender.exchange > 0:
                def_successes = roll_successes(defender.exchange)
                stop_hit_bonus = atk_successes if (STOP_HIT and def_successes >= atk_successes) else 0
                counter_dmg = max(0, def_successes + min(WEAPON_BONUS, defender.exchange) - ARMOR_BONUS + stop_hit_bonus) if def_successes > 0 else 0
                result.attacker_damage_taken = attacker.apply_damage_default(counter_dmg)
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()

    elif atk_maneuver == Maneuver.FEINT:
        attacker.clear_exchange_to_used()
        max_followup = def_commit  # follow-up cannot exceed def commit
                
        if def_maneuver == Maneuver.PARRY:
            defender.clear_exchange_to_used()
            followup = min(atk_followup_commit, attacker.reserve, max_followup)
            if followup > 0:
                attacker.commit(followup)
                followup_successes = roll_successes(followup)
                damage = max(0, followup_successes + min(WEAPON_BONUS, attacker.exchange) - ARMOR_BONUS) if followup_successes > 0 else 0
                result.defender_damage_taken = defender.apply_damage_default(damage)
                attacker.clear_exchange_to_used()

        elif def_maneuver == Maneuver.COUNTER:
            followup = min(atk_followup_commit, attacker.reserve, max_followup)
            attacker.commit(followup)
            def_successes = roll_successes(def_commit)
            total_counter_dmg = max(0, def_successes + min(WEAPON_BONUS, defender.exchange) - ARMOR_BONUS) if def_successes > 0 else 0
            dmg_to_exchange = min(total_counter_dmg, attacker.exchange)
            attacker.exchange -= dmg_to_exchange
            attacker.lost += dmg_to_exchange
            remaining_def_dmg = total_counter_dmg - dmg_to_exchange
            if remaining_def_dmg > 0:
                attacker.apply_damage_default(remaining_def_dmg)
            result.attacker_damage_taken = def_successes
            if attacker.total_hd > 0 and attacker.exchange > 0:
                fu_successes = roll_successes(attacker.exchange)
                dmg = max(0, fu_successes + min(WEAPON_BONUS, attacker.exchange) - ARMOR_BONUS) if fu_successes > 0 else 0
                result.defender_damage_taken = defender.apply_damage_default(dmg)
            attacker.clear_exchange_to_used()
            defender.clear_exchange_to_used()
    
    return result

File: docs/combat.js
import { rollSuccesses } from "./rng.js";

export const Maneuver = Object.freeze({
  PARRY: "Parry",
  COUNTER: "Counter",
  SIMPLE_ATTACK: "SimpleAttack",
  FEINT: "Feint",
  DODGE: "Dodge",
  DECEPTIVE_ATTACK: "DeceptiveAttack",
  DEFENSELESS: "Defenseless",
});

export class CombatantState {
  constructor(maxHd) {
    this.maxHd = maxHd;
    this.reserve = maxHd;
    this.exchange = 0;
    this.used = 0;
    this.lost = 0;
    this.phantom = 0; // temporarily unavailable dice (initiative penalty); drains to lost if hit
  }

  get totalHd() {
    return this.maxHd - this.lost;
  }

  refresh() {
    this.reserve = this.totalHd;
    this.used = 0;
    this.exchange = 0;
    this.phantom = 0;
  }

  commit(dice) {
    dice = Math.min(dice, this.reserve);
    this.reserve -= dice;
    this.exchange += dice;
    return dice;
  }

  clearExchangeToUsed() {
    this.used += this.exchange;
    this.exchange = 0;
  }

  applyDamageDefault(damage) {
    let remaining = damage < 2 ? damage : damage * 3; // TEST
    let applied = 0;
    const fromExchange = Math.min(remaining, this.exchange);
    this.exchange -= fromExchange;
    this.lost += fromExchange;
    remaining -= fromExchange;
    applied += fromExchange;
    if (remaining > 0) {
      const fromReserve = Math.min(remaining, this.reserve);
      this.reserve -= fromReserve;
      this.lost += fromReserve;
      remaining -= fromReserve;
      applied += fromReserve;
    }
    if (remaining > 0) {
      const fromUsed = Math.min(remaining, this.used);
      this.used -= fromUsed;
      this.lost += fromUsed;
      remaining -= fromUsed;
      applied += fromUsed;
    }
    if (remaining > 0) {
      const fromPhantom = Math.min(remaining, this.phantom);
      this.phantom -= fromPhantom;
      this.lost += fromPhantom;
      applied += fromPhantom;
    }
    return applied;
  }
}

export function ExchangeResult() {
  return {
    attackerDamageTaken: 0,
    defenderDamageTaken: 0,
    atkRolled: 0,
    defRolled: 0,
    atkSuccesses: 0,
    defSuccesses: 0,
    followupRolled: 0,
    followupSuccesses: 0,
    dodgeRolled: 0,
    dodgeSuccesses: 0,
    dodgeSucceeded: false,
  };
}

function _evasiveAttack(dodger, target, evasiveCommit, flags) {
  const { evasiveAttack, weaponBonus, armorBonus } = flags;
  if (!evasiveAttack || evasiveCommit <= 0) return 0;
  const committed = dodger.commit(evasiveCommit);
  if (committed <= 0) return 0;
  const successes = rollSuccesses(committed);
  const damage =
    successes > 0
      ? Math.max(0, successes + Math.min(weaponBonus, committed) - armorBonus)
      : 0;
  const applied = target.applyDamageDefault(damage);
  dodger.clearExchangeToUsed();
  return applied;
}

function _resolveDodge(
  attacker,
  defender,
  atkCommit,
  defCommit,
  atkManeuver,
  defManeuver,
  atkFollowupCommit,
  defFollowupCommit,
  atkEvasiveCommit,
  defEvasiveCommit,
  result,
  flags,
) {
  const { armorBonus, weaponBonus } = flags;

  if (atkManeuver === Maneuver.DODGE) attacker.clearExchangeToUsed();
  if (defManeuver === Maneuver.DODGE) defender.clearExchangeToUsed();

  if (
    atkManeuver === Maneuver.DODGE &&
    (defManeuver === Maneuver.DODGE || defManeuver === Maneuver.PARRY)
  ) {
    if (defManeuver === Maneuver.PARRY) defender.clearExchangeToUsed();
    return result;
  }
  if (defManeuver === Maneuver.DODGE && atkManeuver === Maneuver.DODGE) {
    return result;
  }

  // SA / DA (atk) vs Dodge (def)
  if (
    (atkManeuver === Maneuver.SIMPLE_ATTACK ||
      atkManeuver === Maneuver.DECEPTIVE_ATTACK) &&
    defManeuver === Maneuver.DODGE
  ) {
    result.atkRolled = attacker.exchange;
    const atkSuccesses = rollSuccesses(attacker.exchange);
    result.atkSuccesses = atkSuccesses;
    defender.commit(defFollowupCommit);
    result.dodgeRolled = defender.exchange;
    const dodgeRoll = rollSuccesses(defender.exchange);
    result.dodgeSuccesses = dodgeRoll;
    const atkDiceForBonus = attacker.exchange;
    if (dodgeRoll >= Math.max(1, armorBonus)) {
      result.dodgeSucceeded = true;
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
      result.attackerDamageTaken += _evasiveAttack(
        defender,
        attacker,
        defEvasiveCommit,
        flags,
      );
    } else {
      const damage =
        atkSuccesses > 0
          ? Math.max(
              0,
              atkSuccesses +
                Math.min(weaponBonus, atkDiceForBonus) -
                armorBonus,
            )
          : 0;
      result.defenderDamageTaken = defender.applyDamageDefault(damage);
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
    }
    return result;
  }

  // Counter (def) vs Dodge (atk)
  if (atkManeuver === Maneuver.DODGE && defManeuver === Maneuver.COUNTER) {
    attacker.commit(atkFollowupCommit);
    result.dodgeRolled = attacker.exchange;
    const dodgeRoll = rollSuccesses(attacker.exchange);
    result.dodgeSuccesses = dodgeRoll;
    result.defRolled = defCommit;
    const defSuccesses = rollSuccesses(defCommit);
    result.defSuccesses = defSuccesses;
    const defDiceForBonus = defender.exchange;
    if (dodgeRoll >= Math.max(1, armorBonus)) {
      result.dodgeSucceeded = true;
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
      result.defenderDamageTaken += _evasiveAttack(
        attacker,
        defender,
        atkEvasiveCommit,
        flags,
      );
    } else {
      const damage =
        defSuccesses > 0
          ? Math.max(
              0,
              defSuccesses +
                Math.min(weaponBonus, defDiceForBonus) -
                armorBonus,
            )
          : 0;
      result.attackerDamageTaken = attacker.applyDamageDefault(damage);
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
    }
    return result;
  }

  // Feint (atk) vs Dodge (def)
  if (atkManeuver === Maneuver.FEINT && defManeuver === Maneuver.DODGE) {
    attacker.clearExchangeToUsed();
    attacker.commit(atkFollowupCommit);
    result.followupRolled = atkFollowupCommit;
    const followupSuccesses = rollSuccesses(atkFollowupCommit);
    result.followupSuccesses = followupSuccesses;
    defender.commit(defFollowupCommit);
    result.dodgeRolled = defender.exchange;
    const dodgeRoll = rollSuccesses(defender.exchange);
    result.dodgeSuccesses = dodgeRoll;
    const atkDiceForBonus = attacker.exchange;
    if (dodgeRoll >= Math.max(1, armorBonus)) {
      result.dodgeSucceeded = true;
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
      result.attackerDamageTaken += _evasiveAttack(
        defender,
        attacker,
        defEvasiveCommit,
        flags,
      );
    } else {
      const damage =
        followupSuccesses > 0
          ? Math.max(
              0,
              followupSuccesses +
                Math.min(weaponBonus, atkDiceForBonus) -
                armorBonus,
            )
          : 0;
      result.defenderDamageTaken = defender.applyDamageDefault(damage);
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
    }
    return result;
  }

  return result;
}

export function resolveExchange(opts) {
  const {
    attacker,
    defender,
    atkCommit,
    defCommit,
    atkManeuver,
    defManeuver,
    atkFollowupCommit = 0,
    defFollowupCommit = 0,
    atkBonus = 0,
    atkEvasiveCommit = 0,
    defEvasiveCommit = 0,
    weaponBonus = 3,
    armorBonus = 3,
    riposte = false,
    stopHit = false,
    deceptiveAttack = false,
    evasiveAttack = false,
  } = opts;

  const flags = { weaponBonus, armorBonus, evasiveAttack };

  const result = ExchangeResult();
  attacker.commit(atkCommit);
  defender.commit(defCommit);

  let defM =
    defManeuver === Maneuver.DEFENSELESS ? Maneuver.PARRY : defManeuver;

  if (atkManeuver === Maneuver.DECEPTIVE_ATTACK && deceptiveAttack) {
    const bonus = Math.max(
      0,
      Math.min(atkBonus, Math.floor(attacker.reserve / 2)),
    );
    attacker.reserve -= 2 * bonus;
    attacker.exchange += bonus;
    attacker.used += bonus;
  }

  if (atkManeuver === Maneuver.DODGE || defM === Maneuver.DODGE) {
    return _resolveDodge(
      attacker,
      defender,
      atkCommit,
      defCommit,
      atkManeuver,
      defM,
      atkFollowupCommit,
      defFollowupCommit,
      atkEvasiveCommit,
      defEvasiveCommit,
      result,
      flags,
    );
  }

  if (
    atkManeuver === Maneuver.SIMPLE_ATTACK ||
    atkManeuver === Maneuver.DECEPTIVE_ATTACK
  ) {
    const atkRolled = attacker.exchange;
    const atkSuccesses = rollSuccesses(atkRolled);
    result.atkRolled = atkRolled;
    result.atkSuccesses = atkSuccesses;

    if (defM === Maneuver.PARRY) {
      const defRolled = defender.exchange;
      let defSuccesses = rollSuccesses(defRolled) * 2; // TEST;
      result.defRolled = defRolled;
      result.defSuccesses = defSuccesses;
      const damage =
        atkSuccesses > defSuccesses
          ? Math.max(
              0,
              atkSuccesses -
                defSuccesses +
                Math.min(weaponBonus, atkRolled) -
                armorBonus,
            )
          : 0;
      result.defenderDamageTaken = defender.applyDamageDefault(damage);
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
      defSuccesses = Math.floor(defSuccesses / 2); // TEST
      if (riposte && defSuccesses > atkSuccesses) {
        const riposteDamage = Math.max(
          0,
          defSuccesses -
            atkSuccesses +
            Math.min(weaponBonus, defRolled) -
            armorBonus,
        );
        if (riposteDamage > 0) {
          result.attackerDamageTaken =
            attacker.applyDamageDefault(riposteDamage);
        }
      }
    } else if (defM === Maneuver.COUNTER) {
      const totalAtkDmg =
        atkSuccesses > 0
          ? Math.max(
              0,
              atkSuccesses + Math.min(weaponBonus, atkRolled) - armorBonus,
            )
          : 0;
      result.defenderDamageTaken = defender.applyDamageDefault(totalAtkDmg);
      if (defender.totalHd > 0 && defender.exchange > 0) {
        const defRolled = defender.exchange;
        const defSuccesses = rollSuccesses(defRolled);
        result.defRolled = defRolled;
        result.defSuccesses = defSuccesses;
        const stopHitBonus =
          stopHit && defSuccesses >= atkSuccesses ? atkSuccesses : 0;
        const counterDmg =
          defSuccesses > 0
            ? Math.max(
                0,
                defSuccesses +
                  Math.min(weaponBonus, defender.exchange) -
                  armorBonus +
                  stopHitBonus,
              )
            : 0;
        result.attackerDamageTaken = attacker.applyDamageDefault(counterDmg);
      }
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
    }
  } else if (atkManeuver === Maneuver.FEINT) {
    attacker.clearExchangeToUsed();

    if (defM === Maneuver.PARRY) {
      defender.clearExchangeToUsed();
      if (atkFollowupCommit > 0) {
        attacker.commit(atkFollowupCommit);
        result.followupRolled = atkFollowupCommit;
        const followupSuccesses = rollSuccesses(atkFollowupCommit);
        result.followupSuccesses = followupSuccesses;
        const damage =
          followupSuccesses > 0
            ? Math.max(
                0,
                followupSuccesses +
                  Math.min(weaponBonus, attacker.exchange) -
                  armorBonus,
              )
            : 0;
        result.defenderDamageTaken = defender.applyDamageDefault(damage);
        attacker.clearExchangeToUsed();
      }
    } else if (defM === Maneuver.COUNTER) {
      attacker.commit(atkFollowupCommit);
      result.followupRolled = atkFollowupCommit;
      const defRolled = defCommit;
      const defSuccesses = rollSuccesses(defCommit);
      result.defRolled = defRolled;
      result.defSuccesses = defSuccesses;
      const totalCounterDmg =
        defSuccesses > 0
          ? Math.max(
              0,
              defSuccesses +
                Math.min(weaponBonus, defender.exchange) -
                armorBonus,
            )
          : 0;
      result.attackerDamageTaken = attacker.applyDamageDefault(totalCounterDmg);
      if (attacker.totalHd > 0 && attacker.exchange > 0) {
        const fuSuccesses = rollSuccesses(attacker.exchange);
        result.followupSuccesses = fuSuccesses;
        const dmg =
          fuSuccesses > 0
            ? Math.max(
                0,
                fuSuccesses +
                  Math.min(weaponBonus, attacker.exchange) -
                  armorBonus,
              )
            : 0;
        result.defenderDamageTaken = defender.applyDamageDefault(dmg);
      }
      attacker.clearExchangeToUsed();
      defender.clearExchangeToUsed();
    }
  }

  return result;
}
